### ISSUE TITLE
How to persist data?

### ISSUE THREAD

**Issue by lucianozr** [daybrush/moveable#777]
I'm using React-Moveable and I'm having trouble persisting the position and size data I set in the element on canvas.What is the best function to do this? Is onBeforeRender for this?

**Comment by daybrush** [comment]
There is a movable.getRect() method.

is this what you want

**Comment by lucianozr** [comment]
I will use my example. I have an image editor, where the user can add layers and edit them as he wants. But when saving the final image (and I save it in the database) when the user edits the same image again, I need a function to update the moveable so that it adapts to the sizes and positions set previously

**Comment by daybrush** [comment]
@lucianozr 

You can probably store the element's style cssText.
getRect() can simply know the element's 4 points (pos1 ~ 4) and offsetWidth, offsetHeight, left, top information.

**Comment by lucianozr** [comment]
oh, ok... do u have an example of this function? 


I don't think so:
![image](https://user-images.githubusercontent.com/60028283/198699058-b5d54316-a08c-4701-a2c4-906e78ead571.png)

**Comment by daybrush** [comment]
@lucianozr 

use ref(useRef) or access `e.moveable` property in events

```tsx
<Moveable
   ref={moveableRef}
   onBeforeRender={e => {
      // moveableRef.current.getRect();
     e.moveable.getRect();
   }} />
```

**Comment by lucianozr** [comment]
I understand, in getReact I get the positions and sizes to save. But how do I update the moveable the first time it's re-rendered?
Like an initialData, defaultWidth and etc.

**Comment by daybrush** [comment]
@lucianozr 


It has not been released to the public yet, but it can be used as follows.
you can save `moveable.getManager().state` to `firstState`
However, DOM Elements such as target exist. It cannot be stored in stringify or localstorage.

```tsx
<Moveable
     firstRenderState={firstState} />
```

**Comment by lucianozr** [comment]
Oh... okay, i understand. Do i have to pass all the data i got in `e.moveable.getManager().state` for firstRenderState to work? or can i just pass some size and position?

**Comment by daybrush** [comment]
@lucianozr 

It looks like we need to add a feature.
I will consider your function.

The matrix in the state is essential information for rendering.
However, this information can be checked after the first mount.

With only width, height, and pos, it will be difficult to express accurately in the first rendering.

However, it may be inaccurate, but I can make it simpler.
Are you considering SSR?

**Comment by lucianozr** [comment]
no, im not considering the SSR, but im glad you considered my function, im having difficulties with this part of adjusting the moveable on the first launch.

**Comment by daybrush** [comment]
This is probably the first time it will be offered.

persistData can be known through moveable.getRect() information.


```ts
<Moveable
                // target={".target"}
                draggable={true}
                resizable={true}
                rotatable={true}
                viewContainer={document.body}
                persistData={{
                    left:  67.6202,
                    top: 110.6202,
                    offsetHeight: 100,
                    offsetWidth: 100,
                    origin: [130, 173],
                    pos1: [96.7019, 110.6202],
                    pos2: [192.3798, 139.7019],
                    pos3: [67.6202, 206.2981],
                    pos4: [163.2981, 235.3798],
                }}
                onRender={e => {
                    e.target.style.cssText += e.cssText;

                    console.log(e.moveable.getRect());
                }}
            />
```

**Comment by daybrush** [comment]
@lucianozr 

@Raserad 

* `@moveable/helper` 0.0.4
* `lit-moveable` 0.15.0
* `moveable` 0.38.0
* `preact-moveable` 0.40.0
* `react-compat-moveable` 0.26.0
* `react-moveable` 0.41.0
* `svelte-moveable` 0.30.0
* `vue-moveable` 2.0.0-beta.46
* `vue3-moveable` 0.13.0
* `ngx-moveable` 0.35.0


moveable's new version is released.

Use `moveable.getRect()` method (save) and `persistData` prop (restore) .

**Comment by daybrush** [comment]
https://daybrush.com/moveable/storybook2/?path=/story/options--first-render-with-persisted-data

**Comment by lucianozr** [comment]
I am not able to persist the data. I copied your example and in a simple test it is not persisting.

follow the video:

https://drive.google.com/file/d/1Yqrz6reyeLnrMbz4vP1KJCS7y9OLzcjG/view

**Comment by lucianozr** [comment]
I am not able to persist the data. I copied your example and in a simple test it is not persisting.

follow the video:

https://drive.google.com/file/d/19I-YT8c9_xwPQ531RVLP_8fA86CxJhPZ/view

**Comment by daybrush** [comment]
@lucianozr 

The process of setting the controlbox is as follows.

1. render (first) - no shape (render the shape if persistData exists)
2. mount - no shape
3. render- shape (if target exists)
4.updated

If target is not set, controlbox is set to shape with persistData, but
If the target exists, it is set to the shape of the target.

**Comment by keith-ferney** [comment]
@daybrush is there a way I could use the persistData to set the shape of the target back to the original value?

**Comment by daybrush** [comment]
@keith-ferney 

The shape of the target will have to save the css directly.

**Comment by StringKe** [comment]
It might be possible to provide utility functions that convert PersistData to CSS Styles.

`persistDataToCss` convert persist data to css style
`elementToPersisData` convert element to convert persist data

This way, as long as the data from getRect is stored in the

It is possible to set the correct element before the moveable is initialized.
